src/events/broker.py

```diff
@@ -0,0 +1,328 @@
+        if not self.healthy or not self.upgrade.idle:
+            return
+
+        if not self.charm.state.zookeeper.broker_active():
+            self.charm._set_status(Status.ZK_NOT_CONNECTED)
+            return
+
+        # NOTE for situations like IP change and late integration with rack-awareness charm.
+        # If properties have changed, the broker will restart.
+        self.charm.on.config_changed.emit()
+
+        try:
+            if not self.health.machine_configured():
+                self.charm._set_status(Status.SYSCONF_NOT_OPTIMAL)
+                return
+        except SnapError as e:
+            logger.debug(f"Error: {e}")
+            self.charm._set_status(Status.SNAP_NOT_RUNNING)
+            return
+
+        self.charm._set_status(Status.ACTIVE)
+
+    def _on_secret_changed(self, event: SecretChangedEvent) -> None:
+        """Handler for `secret_changed` events."""
+        if not event.secret.label or not self.charm.state.cluster.relation:
+            return
+
+        if event.secret.label == self.charm.state.cluster.data_interface._generate_secret_label(
+            PEER,
+            self.charm.state.cluster.relation.id,
+            "extra",  # pyright: ignore[reportArgumentType] -- Changes with the https://github.com/canonical/data-platform-libs/issues/124
+        ):
+            self.charm.on.config_changed.emit()
+
+    def _on_storage_attached(self, event: StorageAttachedEvent) -> None:
+        """Handler for `storage_attached` events."""
+        # storage-attached usually fires before relation-created/joined
+        if not self.charm.state.peer_relation:
+            event.defer()
+            return
+
+        self.charm.state.unit_broker.update({"storages": self.balancer_manager.storages})
+
+        # new dirs won't be used until topic partitions are assigned to it
+        # either automatically for new topics, or manually for existing
+        # set status only for running services, not on startup
+        self.workload.exec(f"chmod -R 750 {self.workload.paths.data_path}")
+        self.workload.exec(f"chown -R {USER}:{GROUP} {self.workload.paths.data_path}")
+        self.workload.exec(
+            f"""find {self.workload.paths.data_path} -type f -name "meta.properties" -delete || true"""
+        )
+        if self.workload.active():
+            self.charm._set_status(Status.ADDED_STORAGE)
+            # We need the event handler to know about the original event
+            self._on_config_changed(event)
+
+    def _on_storage_detaching(self, _: StorageDetachingEvent) -> None:
+        """Handler for `storage_detaching` events."""
+        # in the case where there may be replication recovery may be possible
+        if self.charm.state.brokers and len(self.charm.state.brokers) > 1:
+            self.charm._set_status(Status.REMOVED_STORAGE)
+        else:
+            self.charm._set_status(Status.REMOVED_STORAGE_NO_REPL)
+
+        self.charm.on.config_changed.emit()
```

Comment:
**todo:** I believe we need to update the `storages` here for the unit, is that right?

Reply:
I wonder about updating here because if the unit is being removed, it does not matter if the unit itself declares 0 storage or if the leader removes the unit from the `broker-capabilities`. If anything else, the current behavior maybe preserves from a wrong `capacityJBOD.json`, if 0 storage for a broker is not allowed. But you are right: if there are several storages on the unit, we do not update the databag at the moment.


Gave it a shot anyway in 89cd5191633bef7b93900e1a5eab24f1b5a1f73f, if that causes issues I'll figure out something else